### PR TITLE
[BREAKING] Refactor testing

### PR DESCRIPTION
**Pull request by Jeonghyun-Kim**
### Major Changes
- use multiprocessing in testing (maxWorkers, with detectOpenHandles)
- add coverage feature to jest configuration
- add ApiBuilder and ApiWrapper (middleware like)
- separate ApiError from previous Error class
- add TestHelper class with testApiHandler static method (which extremely simplifies testing an api handler)
- drop user backend components and api handlers
- remake `connectMongo` utils (with brand new MongoDB class)
- drop `/src/defines/env.ts` (merge it to `/src/utils/env.ts`)
- drop `/src/utils/logger.ts` (will be renewed later? soon? 😁)
- add `sample.env.test.local`
- ensure [`type imports`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export)

### Dependencies Changes
- move 'apidoc' to devDependencies
- add package '@types/node' to devDependencies
- add package 'node-mocks-http' to devDependencies for api testing
- remove `/api/index.ts` and `/api/error.ts` (needs to add other use cases someday..? 🤔)

### Minor Changes
- add /jest/globalConfig.json to .gitignore (which is created by MongoEnvironment.ts while testing)
- ensure 'mongod' process spawns just one time (mongodb-memory-server)
- remove AWS Credentials from `sample.env.local`

### CODE REVIEW

**Review by Jeonghyun-Kim — COMMENTED** [review]
I reviewed by myself once 🧐

File: jest.config.js [hunk 1]
```diff
@@ -13,4 +13,12 @@ module.exports = {
...
   collectCoverage: true,
   collectCoverageFrom: ['<rootDir>/src/{backend,defines,pages}/**/*.ts', '!**/_middleware.ts'],
+  coverageThreshold: {
+    global: {
+      branches: 80,
+      functions: 80,
+      lines: 80,
+      statements: -10,
+    },
+  },
```
Comment (BlueWhaleKo) [inline jest.config.js:23]: impressive :)

File: package.json [hunk 2]
```diff
@@ -43,7 +43,7 @@
     "@types/nprogress": "0.2.0",
     "@types/react": "17.0.19",
     "@types/uuid": "8.3.1",
-    "apidoc": "0.50.1",
+    "apidoc": "0.29.0",
```
Comment (BlueWhaleKo) [inline package.json:45]: Could you explain why you degraded apidoc version?

File: src/backend/test-helper.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      res,
+      statusCode: res._getStatusCode(),
+      jsonData:
+        res.getHeader('Content-Type') === 'application/json'
+          ? (res._getJSONData() as T)
+          : undefined,
+    };
+  }
+}
+
```
Comment (BlueWhaleKo) [inline src/backend/test-helper.ts:32]: nice!

**Review by BlueWhaleKo — APPROVED** [review]
LGTM!

File: package.json (hunk 2 above)
Comment (Jeonghyun-Kim) [reply package.json:45]: because of the apidoc-template compatibility 🤣

**Review by ysyoo11 — APPROVED** [review]
LGTM 👏🏼👏🏼👏🏼

File: src/backend/test-helper.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    handler: NextApiHandler,
+    requestOptions: RequestOptions = {},
+  ): Promise<{
+    req: MockRequest<NextApiRequest>;
+    res: MockResponse<NextApiResponse>;
+    statusCode: number;
+    jsonData?: T;
+  }> {
+    const req = createRequest(requestOptions);
+    const res = createResponse();
```
Comment (kjsik11) [inline src/backend/test-helper.ts:17]: How about declare generic type?

```
  ): Promise<{
    req: MockRequest<NextApiRequest>;
    res: MockResponse<NextApiResponse<T>>;
    statusCode: number;
    jsonData?: T;
  }> {
    const req = createRequest(requestOptions);
    const res = createResponse<NextApiResponse<T>>();

    await handler(req, res);

```

**Review by kjsik11 — APPROVED** [review]
Awesome

File: src/backend/test-helper.ts (hunk 4 above)
Comment (Jeonghyun-Kim) [reply src/backend/test-helper.ts]: resolved by 981ba26

**Pull request merged**